Add `JenkinsBuildSensor`

SasanAhmadi opened the pull request:
The purpose of this PR is to add a new dedicated sensor for jenkins. This contains the sensor file and a test for it.

closes: #22417
related: #22417

SasanAhmadi committed:
add jenkins sensor

SasanAhmadi committed:
add test cases for jenkins sensor

eladkal commented on airflow/providers/jenkins/sensors/jenkins.py:
```diff
@@ -0,0 +1,86 @@
+    Monitor a jenkins job and pass when it is finished building. This is regardless of the build outcome.
+    This sensor depend on python-jenkins library,
+    version >= 0.4.15 to communicate with jenkins server.
+    You'll also need to configure a Jenkins connection in the connections screen.
```
Minimum lib version is 1.0.0 so no need to point that
https://github.com/apache/airflow/blob/fa655fe53dcf34a7b581fe3e881d770921dfc7a0/setup.py#L410

As for the 2nd point also - all sensors work with connections so no reason for special mention of this.

SasanAhmadi replied:
removed the minimum library note
also, removed the obvious fact of connection creation

eladkal commented on airflow/providers/jenkins/sensors/jenkins.py:
```diff
@@ -0,0 +1,86 @@
+                return False
+            else:
+                self.log.info(f"Build #{build_number_to_check} is finished")
+                return True
```
this logic should not be in the sensor code.
The hook should have a general function that returns a build status.
Then the sensor code should be something like:

```
def poke(self, context: 'Context') -> bool: 
    hook = ...
    if hook.get_build_status(...) = 'building'
        return False
    else:
        return True
```

SasanAhmadi replied:
moved the logic to the hook

SasanAhmadi committed:
fix review comment to remove minimum version
it is refelected in setup already
also remove mentioning of the obvious connection need in sensor
https://github.com/apache/airflow/pull/22421/files/33ccbaf0889661f9954fc0ae88b3da32d5ec6d45#r832232224

SasanAhmadi committed:
add dependency note for python-jenkins library

SasanAhmadi committed:
move out get build logic from sensor to hook

eladkal commented on airflow/providers/jenkins/hooks/jenkins.py:
```diff
@@ -46,3 +47,24 @@ def __init__(self, conn_id: str = default_conn_name) -> None:
+            building = build_info['building']
+            return building
+        except jenkins.NotFoundException as err:
+            raise AirflowException(f'Jenkins job status check failed. Final error was: {err.resp.status}')
```
According to the source code of `get_job_info` & `get_build_info` there is no case of raising `NotFoundException`
This is an internal exception of the functions but once caught the functions raise `JenkinsException`

https://opendev.org/jjb/python-jenkins/src/commit/d1e469649574883bc531d35fb3aee207f45c15a5/jenkins/__init__.py#L632

https://opendev.org/jjb/python-jenkins/src/commit/d1e469649574883bc531d35fb3aee207f45c15a5/jenkins/__init__.py#L452

eladkal commented on airflow/providers/jenkins/hooks/jenkins.py:
```diff
@@ -46,3 +47,24 @@ def __init__(self, conn_id: str = default_conn_name) -> None:
         """Get jenkins server"""
         return self.jenkins_server
+
+    def get_build_building_state(self, job_name: str, build_number: int) -> bool:
```
can you add unit test to the hook function?

SasanAhmadi replied:
definitely, I am adding one shortly.

eladkal commented on airflow/providers/jenkins/hooks/jenkins.py:
```diff
@@ -46,3 +47,24 @@ def __init__(self, conn_id: str = default_conn_name) -> None:
+                job_info = self.jenkins_server.get_job_info(self.job_name)
+                build_number_to_check = job_info['lastBuild']['number']
+            else:
+                build_number_to_check = self.build_number
```
There is no `self.build_number` in the hook. I guess this is a mistake?

SasanAhmadi replied:
my bad, fixed

eladkal commented on airflow/providers/jenkins/sensors/jenkins.py:
```diff
@@ -0,0 +1,64 @@
+from airflow.providers.jenkins.hooks.jenkins import JenkinsHook
+from airflow.sensors.base import BaseSensorOperator
+
+class JenkinsSensor(BaseSensorOperator):
```
Should we rename it to `JenkinsBuildSensor`?
This may be more explicit and will be helpful for cases where we will have other sensor in Jenkins (if that make sense)?

SasanAhmadi replied:
That is actually makes a lot of sense, perfect naming for the purpose 
applied

eladkal commented on airflow/providers/jenkins/sensors/jenkins.py:
```diff
@@ -0,0 +1,64 @@
+# under the License.
+
+from typing import Any, Dict, Optional
+from urllib.error import HTTPError, URLError
```
unused imports?

SasanAhmadi replied:
removed

SasanAhmadi committed:
apply fixes for pr review
https://github.com/apache/airflow/pull/22421#discussion_r832546713
https://github.com/apache/airflow/pull/22421#discussion_r832548234

SasanAhmadi committed:
Rename sensor to better suit the purpose
https://github.com/apache/airflow/pull/22421#discussion_r832549698

SasanAhmadi committed:
remove unused import

eladkal commented on tests/providers/jenkins/sensors/test_jenkins.py:
```diff
@@ -0,0 +1,67 @@
+            (3, True,),
+        ]
+    )
+    def test_pole(self, _, build_number, build_state):
```
Did you mean `test_poke`?

SasanAhmadi replied:
typo, fixed

SasanAhmadi replied:
no late night commit ends up good 😄

SasanAhmadi committed:
rename file to match new sensor name
bug fix in test,assert against param instead of fix number

SasanAhmadi committed:
add test for hook's new method

SasanAhmadi committed:
fix typo

SasanAhmadi committed:
remove unnecessary param

SasanAhmadi committed:
fix code review exception, add unit test
fix precommit checks

SasanAhmadi committed:
add optional param type annotation

eladkal commented on airflow/providers/jenkins/sensors/jenkins_build.py:
```diff
@@ -0,0 +1,61 @@
+#
```
The file name should be just `jenkins.py`
So the import will be:

`from airflow.providers.jenkins.sensors.jenkins import JenkinsBuildSensor`

This is the convention (unless special circumstances) for example you can see:
`from airflow.providers.http.sensors.http import HttpSensor`

SasanAhmadi replied:
changed the file name to `jenkins.py`

josh-fell commented on airflow/providers/jenkins/hooks/jenkins.py:
```diff
@@ -46,3 +49,20 @@ def __init__(self, conn_id: str = default_conn_name) -> None:
+            else:
+                build_number_to_check = build_number
+
+            self.log.info(f"Getting build info for {job_name} build number: #{build_number_to_check}")
```
```suggestion
            self.log.info("Getting build info for %s build number: #%s", job_name, build_number_to_check)
```
We shouldn't mix f-strings within logging to ensure the formatting is not done until the very, very last moment when it's needed. You can read this [quick reference](https://dev.to/izabelakowal/what-is-the-best-string-formatting-technique-for-logging-in-python-d1d) for a little insight as to why.

SasanAhmadi replied:
applied
thanks for sharing the blog post

josh-fell commented on airflow/providers/jenkins/sensors/jenkins_build.py:
```diff
@@ -0,0 +1,61 @@
+        self.jenkins_connection_id = jenkins_connection_id
+
+    def poke(self, context: 'Context') -> bool:
+        self.log.info(f"Poking jenkins job {self.job_name}")
```
```suggestion
        self.log.info("Poking jenkins job %s", self.job_name)
```

SasanAhmadi replied:
applied

josh-fell commented on tests/providers/jenkins/sensors/test_jenkins_build.py:
```diff
@@ -0,0 +1,72 @@
+            assert output == (not build_state)
+            assert jenkins_mock.get_job_info.call_count == 0 if build_number else 1
+            assert jenkins_mock.get_build_info.call_count == 1
+            jenkins_mock.get_build_info.assert_called_once_with('a_job_on_jenkins', target_build_number)
```
```suggestion
            jenkins_mock.get_build_info.assert_called_once_with('a_job_on_jenkins', target_build_number)
```
These are redundant in a way and the `assert_called_once_with()` is a better assertion here.

SasanAhmadi replied:
applied

SasanAhmadi committed:
apply @josh-fell comments
https://github.com/apache/airflow/pull/22421/files#r833328675
https://github.com/apache/airflow/pull/22421/files#r833286806
https://github.com/apache/airflow/pull/22421/files#r833329827
https://github.com/apache/airflow/pull/22421/files#r833332407

eladkal commented on tests/providers/jenkins/sensors/test_jenkins_build.py:
```diff
@@ -0,0 +1,71 @@
+#
```
This test file should also be `test_jenkins.py`

SasanAhmadi replied:
fixed

SasanAhmadi committed:
rename test file to match target file

eladkal reviewed (approved):
LGTM

@josh-fell any further comments?

josh-fell reviewed (approved):
LGTM!

eladkal commented on airflow/providers/jenkins/provider.yaml:
```diff
@@ -51,6 +51,11 @@ hooks:
+sensors:
+  - integration-name: Jenkins
+    python-modules:
+      - 'airflow.providers.jenkins.sensors.jenkins_build'
```
```suggestion
      - 'airflow.providers.jenkins.sensors.jenkins'
```

SasanAhmadi replied:
good catch! thanks , applied

eladkal committed:
Update airflow/providers/jenkins/provider.yaml

SasanAhmadi committed:
fix provider file with new name

SasanAhmadi committed:
Merge branch 'jenkins-sensor' of https://github.com/SasanAhmadi/airflow into jenkins-sensor
